### ISSUE TITLE
[Bug] DependencyCRF.log_prob returns a positive value when the input arc score is large.

### ISSUE THREAD

**Issue by wangxinyu0922** [harvardnlp/pytorch-struct#61]
Hi, I found that if I input log_potential into the DependencyCRF with some large values, the marginal distribution will be slightly over 1, and this results in the log_prob returns a very large positive value in training. 
My code is somthing like this:
```
dist = DependencyCRF(arc_scores, lengths=mask.sum(-1))
labels = dist.struct.to_parts(arcs[:,1:], lengths=mask.sum(-1)).type_as(arc_scores)
log_prob = dist.log_prob(labels)
```
I found my `arc_scores` contains a large value like 807687.0625
![图片](https://user-images.githubusercontent.com/17926734/77249023-2c66d200-6c79-11ea-8131-5154d7e6564c.png)
This will result in the marginal contains values like 1.002
![图片](https://user-images.githubusercontent.com/17926734/77249032-4acccd80-6c79-11ea-81bc-9b3ba813ff53.png)
The final `log_prob` will return a very large positive value of 750633.4375. I think this problem is something like floating point precision and can you suggestion something to me?

**Comment by srush** [comment]
Hmm, this is worrisome, we are pretty careful about precision but it is possible we are losing some bits somewhere. I will try to replicate. This is on CUDA right? just for sanity would you mind printing `dist.partition`.

**Comment by wangxinyu0922** [comment]
> 
> 
> Hmm, this is worrisome, we are pretty careful about precision but it is possible we are losing some bits somewhere. I will try to replicate. This is on CUDA right? just for sanity would you mind printing `dist.partition`.

'dist.partition' and the gold tree score:
![图片](https://user-images.githubusercontent.com/17926734/77273598-f5cb9e80-6cee-11ea-9b6d-56981c30ab60.png)

**Comment by srush** [comment]
That is strange. If possible, do you think you could send me a minimal example, perhaps as a colab? Is it plausible that your gold tree is invalid somehow (either non-projective, or with multiple heads?). I am not sure why it would be so different than the partition.

**Comment by wangxinyu0922** [comment]
My dataset contains a tiny proportion of non-projective trees. I think I should try to convert these trees into projective ones.

**Comment by srush** [comment]
I see, so I think there are two things happening here : 
1) minor: the marginal is having a slight underflow, 
2) major: the score is very off because of non-projectivity.

 I would recommend just dropping the non-projective trees, they are not correctly scored by the model and I don't check. If you want to add a check, then I would happy for a PR (it's a non-trivial check to write fast, you need check all arc-pairs).

**Comment by wangxinyu0922** [comment]
> 
> 
> I see, so I think there are two things happening here :
> 
>     1. minor: the marginal is having a slight underflow,
> 
>     2. major: the score is very off because of non-projectivity.
> 
> 
> I would recommend just dropping the non-projective trees, they are not correctly scored by the model and I don't check. If you want to add a check, then I would happy for a PR (it's a non-trivial check to write fast, you need check all arc-pairs).

Thanks for your suggestions. I may try some non-projective trees in the dataset for the model. By the way, can I use `NonProjectiveDependencyCRF` instead of `DependencyCRF`? The document said that `argmax` has not been implemented. Is there any simple way to get the predictions?

**Comment by srush** [comment]
I wouldn't recommend switching between the two. NonProj will be faster and more general, but less accurate in cases where the languages 99% projective. Unfortunately there is currently no easy way to get an argmax tree! You can try taking the max for each arc, but that will not guarantee it is a tree. 

If you are interested in implementing the argmax, the algorithm is the Chiu-Liu-Edmonds algorithm. It shouldn't be too hard to implement, it's just a bit different so I didn't include it.

**Comment by wangxinyu0922** [comment]
Now I'm trying to reproduce the problem with the following code:
```
arc_scores[:,:,:]=500000
arc_scores[:,:,:25]=-500000
dist = DependencyCRF(arc_scores, lengths=mask.sum(-1))
labels = dist.struct.to_parts(arcs[:,1:], lengths=mask.sum(-1)).type_as(arc_scores)
log_prob = dist.log_prob(labels)
```
Such way will make the `dist.marginals` becomes larger than 1.
![图片](https://user-images.githubusercontent.com/17926734/77282519-20c0ed00-6d05-11ea-9d41-b6bf85e6273a.png)

I'm still trying to make the log_prob become a positive one.

**Comment by wangxinyu0922** [comment]
Well, now I'm trying cases for `NonProjectiveDependencyCRF `. The code is like this:
```
arc_scores[:,:,:]=0
dist = NonProjectiveDependencyCRF(arc_scores, lengths=mask.sum(-1))
labels = dist.struct.to_parts(arcs[:,1:], lengths=mask.sum(-1)).type_as(arc_scores)
log_prob = dist.log_prob(labels)
```

When the sentence length is long (I tried 50 here), the output `log_prob` is `-inf` and the `dist.partition` is `inf`. When sentence length is shorter (I tried one with 9 words), the model returns a correct value. Do you have any ideas?

**Comment by srush** [comment]
No ideas yet, but I will look into both issues.

**Comment by kmkurn** [comment]
Hi, sorry for jumping in, but the issue with `NonProjectiveDependencyCRF` is I believe the line below:
https://github.com/harvardnlp/pytorch-struct/blob/9f5a0f5b5a41de9eb64b6abf23246677ce5a02ae/torch_struct/deptree.py#L196

It should've used `tensor.logdet` to avoid numerical error for large matrices.

**Comment by wangxinyu0922** [comment]
After setting a breakpoint to find out when the problem occurs in my training code, I find that probably there is something wrong with the `length` argument in `DependencyCRF` including the problem of the underflow. The code is:

```
temp_arcs=torch.load('temp_arcs.pt')
temp_mask=torch.load('temp_mask.pt')
# (batch, sent_len+1, sent_len+1), shape: (dependency, head)
# -1e-9 is a masked value in my code
temp_scores=torch.load('temp_scores.pt')
# pdb.set_trace()
# basic
dist=generate_tree(temp_scores,temp_mask,is_mst=False)
labels = dist.struct.to_parts(temp_arcs, lengths=temp_mask.sum(-1)).type_as(temp_scores)
log_prob = dist.log_prob(labels)
# tensor([3.0518e-05], device='cuda:0', grad_fn=<SubBackward0>)
print(log_prob)

# masking the dependency part
temp_scores2=temp_scores*temp_mask.unsqueeze(-1)
dist=generate_tree(temp_scores2,temp_mask,is_mst=False)
labels = dist.struct.to_parts(temp_arcs, lengths=temp_mask.sum(-1)).type_as(temp_scores2)
log_prob = dist.log_prob(labels)
# tensor([3.0518e-05], device='cuda:0', grad_fn=<SubBackward0>)
print(log_prob)

# masking the head part
temp_scores3=temp_scores*temp_mask.unsqueeze(-2)
dist=generate_tree(temp_scores3,temp_mask,is_mst=False)
labels = dist.struct.to_parts(temp_arcs, lengths=temp_mask.sum(-1)).type_as(temp_scores3)
log_prob = dist.log_prob(labels)
# tensor([-0.0044], device='cuda:0', grad_fn=<SubBackward0>)
print(log_prob)
```

I have uploaded these files on the [OneDrive](https://1drv.ms/u/s!Am53YNAPSsodg7lbYWx1799Hpntvng?e=unstor), I hope this will help us to solve the problem quickly.

**Comment by srush** [comment]
@kmkurn ah fantastic, I changed that line. Can you make another issue about variable length?

**Comment by srush** [comment]
@wangxinyu0922 thanks for the repro, I will see what I can do to fix.

**Comment by srush** [comment]
@wangxinyu0922 I think your code had a bug in it. 

![image](https://user-images.githubusercontent.com/35882/77460658-c775be00-6dd7-11ea-9984-a7e734cc4d45.png)

Your mask is zeroing out the potentials for the 0th head word.

**Comment by wangxinyu0922** [comment]
> 
> 
> @wangxinyu0922 I think your code had a bug in it.
> 
> ![image](https://user-images.githubusercontent.com/35882/77460658-c775be00-6dd7-11ea-9984-a7e734cc4d45.png)
> 
> Your mask is zeroing out the potentials for the 0th head word.

In my code, there is an redundant '<ROOT>' word at position 0, and the `temp_score` is a shape of `(dependency, head)`. Therefore masking 0th word is correct. To fit the requirements of `DependencyCRF`, the score of 0th head word should be on the diagonal of the `log_potential` matrix without the '<ROOT>' and the matrix should be the shape of  (head, dependency), which is processed in the `process_potential` function. So this should not be the problem.

Why the third example (which masks the score of '<ROOT>' to 0 for all words/setting the diagonal of `log_potential` to be 0) gets a correct negative `log_prob` while other does not? Do I missed something else?

**Comment by srush** [comment]
Sorry I'm still having trouble understanding. I'm really trying to figure it out, but I need a more clear example.

1) The 0 word is a real word in the model, if you change it's scores it will impact the model partition. If you want to change the root you need to change the diagonal.
2) Making a log potential 0 does not remove it from the model. If you want to mask out a decision completely you need to make it -inf.

**Comment by wangxinyu0922** [comment]
@srush 

I think I have figured out why this problems happens:
1. The computation of gold tree score does not takes `lengths` (the mask) into consideration.
2. The tiny value `-1e9` in my code results in underflow in the calculation.

When taking your suggestion that mask out the log potential with -inf, I found that the `log_prob` becomes `nan`. I digged deeper and find out that the `v` in the `dist.log_prob` function in `distribution.py` does not consider `lengths` but `dist.partition` takes, which makes the calculation of `log_prob` is a wrong score. 
[](https://github.com/harvardnlp/pytorch-struct/blob/a8ed72be1a0c49bc67351803d5112e5577d4180f/torch_struct/distributions.py#L62)

And the calculation of `score` in `helper.py` use `mul` I think that makes the results underflow:
[](https://github.com/harvardnlp/pytorch-struct/blob/a8ed72be1a0c49bc67351803d5112e5577d4180f/torch_struct/helpers.py#L72)

**Comment by srush** [comment]
I still do not understand exactly where you think the problem is. Do you think the problem is with the score function? We could try use torch.where  instead of torch.mul and see if the value is different. 

If you think the partition calculation is the issue, can you please send me a working minimal example where the partition is less than the value of a single tree?

**Comment by wangxinyu0922** [comment]
[debug.zip](https://github.com/harvardnlp/pytorch-struct/files/4381554/debug.zip)
I tried to use torch.where instead of torch.mul and the values are the same. Here is an example that the partition is less than the value of a single tree

**Comment by srush** [comment]
Thanks this is a good repro case. I see the issue, but I do not yet have a good fix. Can't tell if it is an inherent issue in the algorithm, or a bug somewhere. 

![image](https://user-images.githubusercontent.com/35882/77866524-7a438300-7201-11ea-8f7b-09322903118b.png)
